### PR TITLE
Documentation refactor

### LINKED ISSUE #348: Update sphinx document for all the new models
Since new models are added to Anomalib, this needs to be reflected in the sphinx documentation as well. To do this we have to update a couple of files.

- [ ] models.rst
- [ ] Papers.rst
- [ ] benchmark.md
- [ ] Add metrics to main README

### PR DESCRIPTION
**Pull request by samet-akcay**
# Description

- Provide a summary of the modification as well as the issue that has been resolved. List any dependencies that this modification necessitates.
- Fixes #348 #304 #543 

image

## Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which refactors the code base)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist

- [ ] My code follows the [pre-commit style and check guidelines](https://openvinotoolkit.github.io/anomalib/guides/using_pre_commit.html#pre-commit-hooks) of this project.
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests pass locally with my changes

### CODE REVIEW

**Comment by jpcbertoldo** [comment]
maybe #544 could be integrated here?
(tbh i didnt read it all so if it's already in there just ignore the comment)

**Review by ashwinvaidya17 — APPROVED** [review]
Thanks for the efforts!

File: docs/source/how_to_guides/adding_a_new_model.rst [hunk 1]
```diff
@@ -0,0 +1,198 @@
+Adding a New Model to Anomalib
+==============================
+New models can be added to anomalib and are always welcome. To add new models, the following steps should be followed:
```
Comment (djdameln) [inline docs/source/how_to_guides/adding_a_new_model.rst:3]: > New models can be added to anomalib and are always welcome

Maybe this could be rephrased. To prevent cluttering the library, I think we would only be interested in adding a model to the main branch if it's accompanied by a published paper. Or if we've tested it and decide it solves a specific problem or fills a gap in the library.

File: docs/source/how_to_guides/adding_a_new_model.rst [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    # PL Trainer Args. Don't add extra parameter here.
+    trainer:
+        accelerator: auto # <"cpu", "gpu", "tpu", "ipu", "hpu", "auto">
+        ...
+
+
+
+Create a ``torch_model.py`` Module
+----------------------------------
+This file contains the torch model implementation that inherits from ``torch.nn.Module`` and performs a basic forward-pass. The advantage of storing the model in a separate ``torch_model.py`` file is that the model is de-coupled from the rest of the anomalib implementations and could be used outside the library as well. Basic implementation would look like as follows:
```
Comment (djdameln) [inline docs/source/how_to_guides/adding_a_new_model.rst:62]: > and performs a basic forward-pass

"defines the model architecture and performs a basic forward-pass"?

File: docs/source/how_to_guides/adding_a_new_model.rst [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        """New Model Module."""
+        def __init__(self):
+            pass
+        def forward(self, x):
+            pass
+
+
+Create a ``lightning_model.py`` Module
+--------------------------------------
+This module contains the actual lightning model implementation that inherits from the ``AnomalModule``, which already has the ``anomalib`` related attributes and methods. The user does not need to worry about the boilerplate code and only needs to implement the training and validation logic of the algorithm.
```
Comment (djdameln) [inline docs/source/how_to_guides/adding_a_new_model.rst:76]: ~~actual~~

File: docs/source/how_to_guides/adding_a_new_model.rst [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+    class NewModelLoss(nn.Module):
+        """NewModel Loss."""
+
+        def forward(self) -> Tensor:
+            """Calculate the NewModel loss."""
+            pass
+
+Create an ``anomaly_map.py`` File - [Optional]
+---------------------------------------------
+Similar to the loss.py module, the anomaly map.py module is optional depending on the capabilities of the algorithm. This module should be implemented if the algorithm supports segmentation so that the location of the anomaly can be predicted pixel-by-pixel.
```
Comment (djdameln) [inline docs/source/how_to_guides/adding_a_new_model.rst:106]: ~~should~~ may

File: docs/source/how_to_guides/adding_a_new_model.rst [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            pass
+        def training_step(self, batch):
+            pass
+        ...
+        def validation_step(self, batch):
+            pass
+
+Create a ``loss.py`` File - [Optional]
+--------------------------------------
+This module's availability is dependent on the algorithm. If the algorithm requires a custom loss function, this file contains the subclass of the torch.nn.Module class implementation. This loss would subsequently be utilized by the lightning module.
```
Comment (djdameln) [inline docs/source/how_to_guides/adding_a_new_model.rst:93]: ~~this file contains the~~ this file may contain the

If the custom loss function is very simple, it might not be needed to have a separate class for this.

**Review by djdameln — CHANGES_REQUESTED** [review]
Minor comments

**Comment by samet-akcay** [comment]
> Minor comments

@djdameln, thanks for the review. I've addressed all of your comments.

**Review by djdameln — APPROVED** [review]
Thanks!
